File: scala/src/main/scala/string/LongestPalindrome.scala
package org.whsv26.leetcode
package string

object LongestPalindrome extends App {

  def longestPalindrome(s: String): Int = {
    val occurencesByChar = s.groupMapReduce(identity)(_ => 1)(_ + _)

    val occurenceSum = occurencesByChar.values.collect {
      case n if n % 2 == 0 => n
      case n if n % 2 != 0 => n - 1
    }.sum

    if (occurenceSum < s.length) occurenceSum + 1
    else occurenceSum
  }

  assert(11 == longestPalindrome("aaabbbccddee"))
  assert(11 == longestPalindrome("aaabbccddee"))
  assert(7 == longestPalindrome("xqaaaddee"))
  assert(1 == longestPalindrome("abcde"))

}

File: java-problems/src/main/java/org/whsv26/leetcode/LongestPalindrome.java
package org.whsv26.leetcode;

import java.util.HashMap;

public class LongestPalindrome {

    public static void main(String[] args) {
        var solutions = new Solution();
        assert (11 == solutions.longestPalindrome("aaabbbccddee"));
        assert (11 == solutions.longestPalindrome("aaabbccddee"));
        assert (7 == solutions.longestPalindrome("xqaaaddee"));
        assert (1 == solutions.longestPalindrome("abcde"));
    }

    static class Solution {
        public int longestPalindrome(String s) {
            var occurByChar = new HashMap<Integer, Integer>();
            s.chars().forEach(c -> occurByChar.merge(c, 1, Integer::sum));

            var occurSum = occurByChar.values().stream()
                .mapToInt(n -> n % 2 == 0 ? n : n - 1)
                .sum();

            return occurSum < s.length() ? occurSum + 1 : occurSum;
        }
    }
}